SwiftLeeds/Network/URLSession.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  URLSession.swift
+//  SwiftLeeds
+//
+//  Created by Matthew Gallagher on 24/09/2023.
+//
+
+import Foundation
+
+public extension URLSession {
+    static var awaitConnectivity: URLSession = {
+        let configuration = URLSessionConfiguration.default
+        configuration.waitsForConnectivity = true
+        configuration.timeoutIntervalForRequest = 5 //60
+        configuration.timeoutIntervalForResource = 5 //60 * 30
+        configuration.urlCache = nil
+        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
+        return URLSession(configuration: configuration)
+    }()
+
+    func cached<Response: Decodable>(_ request: Request<Response>, using decoder: JSONDecoder = .init(),
+                                     dateDecodingStrategy: JSONDecoder.DateDecodingStrategy = .deferredToDate,
+                                     fileManager: FileManager = .default, filename: String? = nil) async throws -> Response {
+        let filename = filename ?? request.url.lastPathComponent
+        let path = fileManager.temporaryDirectory.appendingPathComponent(filename)
+
+        guard let data = fileManager.contents(atPath: path.path.appending(".json")) else { throw NetworkError.cacheNotFound }
+
+        let decoded = Task.detached(priority: .userInitiated) {
+            try Task.checkCancellation()
+            decoder.dateDecodingStrategy = dateDecodingStrategy
+            return try decoder.decode(Response.self, from: data)
+        }
+
+        return try await decoded.value
+    }
+
+    func decode<Response: Decodable>(_ request: Request<Response>, using decoder: JSONDecoder = .init(),
+                                     dateDecodingStrategy: JSONDecoder.DateDecodingStrategy?,
+                                     fileManager: FileManager = .default, filename: String? = nil) async throws -> Response {
+        let filename = filename ?? request.url.lastPathComponent
+        let path = fileManager.temporaryDirectory.appendingPathComponent("\(filename).json")
+
+        let decoded = Task.detached(priority: .userInitiated) {
+            do {
+                let (data, response) = try await self.data(for: request.urlRequest)
+
+                try Task.checkCancellation()
+
+                guard let response = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
+
+                if let eTagKey = request.eTagKey, let eTagValue = response.value(forHTTPHeaderField: "Etag") {
+                    UserDefaults.standard.set(eTagValue, forKey: eTagKey)
+                }
```

Comment:
I think we should move the eTag capture to after the `data.write` call.

If we fail to save it, then we'd want to request the data back from origin without eTag, right?

Reply:
Good suggestion

Reply:
If we capture eTag before we populate cache then there's a journey there which means cache fails, etag populates, on refresh we use eTag, it tells us no new content and to use cache, but cache failed so nothing is ever shown.

eTag must only be used when cache is there as a backup.